korge/src/commonMain/kotlin/com/soywiz/korge3d/animation/PredefinedPlaybackPatterns.kt

```diff
@@ -0,0 +1,25 @@
+package com.soywiz.korge3d.animation
+
+import kotlin.math.PI
+import kotlin.math.sin
+
+val repeatInfinitely = PlaybackPattern { time: NormalizedTime -> NormalizedProgress(time.value % 1.0) }
+
+fun constant(progress: Double) = PlaybackPattern { NormalizedProgress(progress) }
+
+val oscillation = PlaybackPattern { x -> NormalizedProgress(sin(x.value * 2.0 * PI) * 0.5 + 0.5) }
+
+fun speed(speed: Double) = PlaybackModifier(
+    argumentModifier = { time: NormalizedTime -> NormalizedTime(time.value * speed) })
+
+fun fadeOut(periods: Double): PlaybackModifier {
+    require(periods > 0.0) { "Periods must be positive, and $periods given!" }
+    return PlaybackModifier(
+        valueModifier = { x, y ->
+            NormalizedProgress(when {
+                x.value < periods -> y.value * (-x.value/periods + 1)
+                else -> 0.0
+            })
+        }
+    )
+}
```

Comment:
I have a few concerns about this. Let me elaborate:

As I can see on the tests this reads well, but I think it won't be easy to discover how to use the API without looking at the samples, since this is not an enum. So from a PlaybackPattern one could find the with function by pressing the dot key, but then what goes next? A PlaybackModifier that is an inline class. It is not obvious there are a few global functions doing that. I believe that even if it is not that elegant, providing `PlaybackPattern.withSpeed` and `PlaybackPattern.fadeOut` would be easier to discover to people using the API.

The other concern I have is that usually in the IDE, when writing speed, the autocompletion would display this function despite of the context. One could see a `speed` function returning a `PlaybackModifier` and be a bit confused because of the lack of context.

Regarding to `repeatInfinitely` and `oscillation` I have a similar concern. What about defining them with `PlaybackPattern.Companion.repeatInfinitely`. As said, not that elegant, but people could try to check if there are `static` properties associated to the class they have to provide. Hopefully in the future Kotlin will support writing enum members without having to put the full enum name and a way to do something similar with inline classes.

Additionally as stated earlier, can you have a look to the Easing interface and figure out if we could merge/reuse/extend them? This funcionality looks more generic to the korge3d.animation package, and at least conceptually looks pretty similar to Easing. Maybe providing a way to generate Easings and then modiying them with modifiers.

Reply:
Right, I get your concerns. I planned to just document all existing predefined patterns along with plots depicting them on KorGE website, but I agree that we should try to make the code as self-descriptive and discoverable as possible. That's why I'd take a step back and in this PR focus on adding top-level API (essentially: `Animation3D` takes `PlaybackPattern` as an argument), and take care of `PlaybackModifier` and easings (see my response to the other comment) in separate PR(s). After this simple change, the user can already define arbitrarily complex patterns, and whatever comes in next PRs will be simplifying user's lives.

The only thing I'll have to add in this PR is `repeatInfinitely` which I can surely make a member of `PlaybackPattern`'s companion object :)

Reply:
We have already a few functions in DSLs that are not obvious to some people since they are available inside several levels of DSL blocks. So whenever possible we want to keep all the scopes as clean of possible in terms of unrelated exposed symbols that are available.

It is a bit sad because looking at the tests, those reads pretty well and it is elegant, but we also want to be nice in terms of discovery and clean scopes.

In the KorGE engine core we also had to make some tradeoffs of clean and elegant stuff to prevent allocations, specially with vector operations like: `a.x + b.x; a.y + b.y` instead of just `a + b`. Hopefully we will be able to simplify that when the JVM releases project Valhalla as long as there are conterparts at Kotlin/Native and hopefully with JS output tricks (or WebAssembly when it supports GC and exceptions).